File: internal/telemetry/model/backup.go
package model

// The backup document: everything guard was configured with, as one file.
//
// A backup is guard's *configuration* — the machines, the commands, the cloud
// accounts, the saved views, the alert rules, the secrets, the members and the
// stored environment — and never its telemetry. Logs, traces and metrics are the
// part that grows without bound and the part that is reproduced by the next
// minute of ingest; a backup carrying them would be a database copy wearing a
// different extension, and nobody would take one.
//
// The shape is a table per section rather than a struct per feature, and that is
// deliberate: a domain-shaped document has to be edited every time a column is
// added, and the first time somebody forgets, a restore quietly drops the column
// they just introduced. Here the columns come from the database at both ends and
// are matched by name, so a backup taken by an older guard restores into a newer
// one with the new columns left at their defaults, and a column guard no longer
// has is reported rather than fatal.

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"math"
	"strconv"
	"strings"
)

// BackupFormat is the document version. It is written by the exporter and
// checked by the importer, which refuses anything higher: a file from a newer
// guard may say things this one would silently ignore.
const BackupFormat = 1

// How the sealed values travelled.
const (
	// BackupSecretsPlain: the credentials are in the file as themselves. This
	// is what a backup with no passphrase is, and it is the only way a restore
	// can be a restore — an instance whose SSH passwords and API keys all came
	// back empty is not the instance the file described, and the dashboard's
	// way of saying so is a page of "no stored key" three weeks later.
	//
	// The consequence is the whole of the warning on the page: **the file is
	// the credentials**. Anybody holding it holds them, which is exactly what
	// the passphrase below is for.
	BackupSecretsPlain = "plaintext"
	// BackupSecretsPassphrase: sealed values were opened with this instance's
	// key and re-sealed under a key derived from a passphrase. The file is
	// portable to any guard, and worth nothing without the passphrase.
	BackupSecretsPassphrase = "passphrase"
	// BackupSecretsOmitted is only ever read, never written: the first version
	// of this left the credentials out, and those files still restore — with
	// every credential blank, which is what they actually carry.
	BackupSecretsOmitted = "omitted"
)

// Backup is the whole file.
type Backup struct {
	Format    int    `json:"format"`
	Guard     string `json:"guard_version"`
	CreatedNs int64  `json:"created_ns"`
	// Secrets is one of the two constants above.
	Secrets string `json:"secrets"`
	// KDF is present exactly when Secrets is BackupSecretsPassphrase.
	KDF *BackupKDF `json:"kdf,omitempty"`
	// Notes is anything about this file somebody has to know, in words —
	// today, a value that could not be read on the way out. A file quietly
	// missing the section somebody most wanted is the failure this exists to
	// prevent.
	Notes  []string      `json:"notes,omitempty"`
	Tables []BackupTable `json:"tables"`
}

// BackupKDF is how the passphrase becomes a key, written down so the importer
// does not have to agree with the exporter by convention.
type BackupKDF struct {
	Algorithm string `json:"algorithm"`
	Salt      string `json:"salt"`
	N         int    `json:"n"`
	R         int    `json:"r"`
	P         int    `json:"p"`
	// Check is a known plaintext sealed with the derived key. It is the reason
	// a wrong passphrase is a sentence on the page rather than a restore that
	// half-works: it is opened before anything is written, and before the file
	// is even inspected for values that might happen to decrypt.
	Check string `json:"check"`
}

// BackupTable is one table, with its columns named once and its rows as arrays.
// Named columns and positional rows, because a backup of a thousand secrets
// should not repeat every column name a thousand times.
type BackupTable struct {
	Name    string          `json:"name"`
	Label   string          `json:"label"`
	Group   string          `json:"group"`
	Columns []string        `json:"columns"`
	Rows    [][]BackupValue `json:"rows"`
}

// BackupKind is the SQLite storage class of one value.
type BackupKind uint8

const (
	BackupNull BackupKind = iota
	BackupInt
	BackupReal
	BackupText
	BackupBlob
)

// BackupValue is one cell, typed.
//
// It carries its own JSON encoding for two reasons, both of which were bugs in
// the obvious version. A timestamp in nanoseconds does not survive a round trip
// through a JSON number decoded as float64 — 1.7e18 is past the point where
// doubles count in ones — so integers are parsed from their literal text. And a
// BLOB decoded into `any` comes back as a base64 *string*, indistinguishable
// from a text column that happens to hold base64, so blobs are wrapped in an
// object and stay blobs.
type BackupValue struct {
	Kind BackupKind
	Int  int64
	Real float64
	Text string
	Blob []byte
}

// The five constructors, so nothing has to set Kind by hand.
func BackupNullValue() BackupValue          { return BackupValue{Kind: BackupNull} }
func BackupIntValue(v int64) BackupValue    { return BackupValue{Kind: BackupInt, Int: v} }
func BackupRealValue(v float64) BackupValue { return BackupValue{Kind: BackupReal, Real: v} }
func BackupTextValue(v string) BackupValue  { return BackupValue{Kind: BackupText, Text: v} }

// BackupBlobValue keeps nil distinct from empty: a NULL password and a
// zero-length one are different rows, and the difference is "never set" versus
// "set to nothing".
func BackupBlobValue(v []byte) BackupValue {
	if v == nil {
		return BackupValue{Kind: BackupNull}
	}
	return BackupValue{Kind: BackupBlob, Blob: v}
}

// Any is the value as database/sql wants it.
func (v BackupValue) Any() any {
	switch v.Kind {
	case BackupInt:
		return v.Int
	case BackupReal:
		return v.Real
	case BackupText:
		return v.Text
	case BackupBlob:
		return v.Blob
	default:
		return nil
	}
}

// IsNull answers the one question the sealed-column handling asks.
func (v BackupValue) IsNull() bool { return v.Kind == BackupNull }

func (v BackupValue) MarshalJSON() ([]byte, error) {
	switch v.Kind {
	case BackupInt:
		return []byte(strconv.FormatInt(v.Int, 10)), nil
	case BackupReal:
		if math.IsInf(v.Real, 0) || math.IsNaN(v.Real) {
			// Neither is representable in JSON, and neither is a number any
			// column here should hold; null is the honest answer.
			return []byte("null"), nil
		}
		text := strconv.FormatFloat(v.Real, 'g', -1, 64)
		// A REAL that happens to be whole formats as "2", which would read back
		// as an integer. The decimal point is what keeps the storage class.
		if !strings.ContainsAny(text, ".eE") {
			text += ".0"
		}
		return []byte(text), nil
	case BackupText:
		return json.Marshal(v.Text)
	case BackupBlob:
		return json.Marshal(struct {
			B64 string `json:"b64"`
		}{base64.StdEncoding.EncodeToString(v.Blob)})
	default:
		return []byte("null"), nil
	}
}

func (v *BackupValue) UnmarshalJSON(raw []byte) error {
	raw = bytes.TrimSpace(raw)
	if len(raw) == 0 {
		return errors.New("backup: empty value")
	}
	switch raw[0] {
	case 'n':
		*v = BackupValue{Kind: BackupNull}
		return nil
	case '"':
		var text string
		if err := json.Unmarshal(raw, &text); err != nil {
			return err
		}
		*v = BackupTextValue(text)
		return nil
	case '{':
		var wrapper struct {
			B64 string `json:"b64"`
		}
		if err := json.Unmarshal(raw, &wrapper); err != nil {
			return err
		}
		blob, err := base64.StdEncoding.DecodeString(wrapper.B64)
		if err != nil {
			return err
		}
		if blob == nil {
			blob = []byte{}
		}
		*v = BackupValue{Kind: BackupBlob, Blob: blob}
		return nil
	case 't', 'f':
		// Nothing guard writes is a JSON boolean, but a file edited by hand
		// might be, and 1/0 is what SQLite would have stored anyway.
		var flag bool
		if err := json.Unmarshal(raw, &flag); err != nil {
			return err
		}
		if flag {
			*v = BackupIntValue(1)
		} else {
			*v = BackupIntValue(0)
		}
		return nil
	}
	text := string(raw)
	if !strings.ContainsAny(text, ".eE") {
		if number, err := strconv.ParseInt(text, 10, 64); err == nil {
			*v = BackupIntValue(number)
			return nil
		}
	}
	number, err := strconv.ParseFloat(text, 64)
	if err != nil {
		return errors.New("backup: " + text + " is not a value")
	}
	*v = BackupRealValue(number)
	return nil
}

// BackupSummary is what the page draws before anybody presses anything: what a
// backup taken now would contain, section by section.
type BackupSummary struct {
	Format int                  `json:"format"`
	Guard  string               `json:"guard_version"`
	Tables []BackupTableSummary `json:"tables"`
	Sealed int                  `json:"sealed"`
	// Excluded names the tables a backup leaves behind, so the page can say so
	// rather than leaving somebody to wonder whether their logs are in there.
	Excluded []string `json:"excluded"`
}

// BackupTableSummary is one section's line on that page.
type BackupTableSummary struct {
	Name   string `json:"name"`
	Label  string `json:"label"`
	Group  string `json:"group"`
	Rows   int    `json:"rows"`
	Sealed int    `json:"sealed"`
}

// RestoreReport is what came back from a restore. Every table is named with
// what it took, because "restored" with no numbers is a word rather than an
// answer.
type RestoreReport struct {
	Guard     string         `json:"guard_version"`
	CreatedNs int64          `json:"created_ns"`
	Tables    []RestoreTable `json:"tables"`
	Rows      int            `json:"rows"`
	// Sealed is how many credentials were re-sealed with this instance's key,
	// and Blank how many came back empty because the file carried none.
	Sealed   int      `json:"sealed"`
	Blank    int      `json:"blank"`
	Warnings []string `json:"warnings"`
	// Restart says the file carried stored configuration, which a running
	// process took its environment from at startup and will not re-read.
	Restart bool `json:"restart"`
}

// RestoreTable is one table's share of that.
type RestoreTable struct {
	Name    string   `json:"name"`
	Label   string   `json:"label"`
	Group   string   `json:"group"`
	Rows    int      `json:"rows"`
	Skipped []string `json:"skipped,omitempty"`
}

File: client/public/backup.js
// The backup page: what a backup would hold, the download, and the restore.
//
// The counts come from GET /api/backup rather than being written into the page,
// so the day somebody adds a table to the catalogue the line says so without
// being edited.
//
// The file itself never goes through the store. Everything else on the
// dashboard is a value two pages might share; this is a download somebody asked
// for once, and caching it would mean a backup taken from a page that had been
// open since Tuesday.
import { adminHeaders, el, muted, number, qs, request } from "./core.js";
import { ensure, forgetAll } from "./store.js";
import { ask } from "./cluster.js";

function status(message) {
  const node = qs("[data-backup-status]");
  if (node) node.textContent = message;
}

function restoreStatus(message) {
  const node = qs("[data-backup-restore-status]");
  if (node) node.textContent = message;
}

// One line, not a list. What a person decides from here is whether the file is
// worth taking — how much configuration there is and how many credentials are
// in it — and an export is all of it or none of it, so a table of sections
// reads like a chooser that does not exist.
function drawSummary(summary) {
  const host = qs("[data-backup-sections]");
  if (!host) return;
  const rows = (summary.tables || []).reduce((total, table) => total + table.rows, 0);
  const sections = (summary.tables || []).filter((table) => table.rows > 0).length;
  const sealed = summary.sealed || 0;
  host.textContent = `${number.format(rows)} row${rows === 1 ? "" : "s"} across ` +
    `${sections} section${sections === 1 ? "" : "s"}, including ${number.format(sealed)} ` +
    `stored credential${sealed === 1 ? "" : "s"} — sealed under your passphrase, or written plainly without one.`;
}

export async function refreshBackup() {
  const host = qs("[data-backup-sections]");
  if (!host) return;
  try {
    await ensure("backup", () => request("/api/backup", { headers: adminHeaders() }), drawSummary);
  } catch (failure) {
    host.textContent = failure.message;
  }
}

// The download. A blob and a synthetic click, because the file comes back from
// a POST carrying a passphrase — a plain link would have to put it in a URL,
// where it would land in the access log of every proxy in front of guard.
function save(doc) {
  const stamp = new Date().toISOString().slice(0, 19).replaceAll(":", "-");
  const blob = new Blob([JSON.stringify(doc, null, 2)], { type: "application/json" });
  const url = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = url;
  link.download = `guard-backup-${stamp}.json`;
  document.body.append(link);
  link.click();
  link.remove();
  URL.revokeObjectURL(url);
}

async function exportBackup() {
  const passphrase = qs("[data-backup-passphrase]")?.value || "";
  status("Reading the configuration…");
  try {
    const doc = await request("/api/backup", {
      method: "POST", headers: adminHeaders(),
      body: JSON.stringify({ passphrase }),
    });
    save(doc);
    const rows = (doc.tables || []).reduce((total, table) => total + table.rows.length, 0);
    // Anything the export could not carry travels with it as a note, and is
    // said here rather than left in the file — today that is the stored
    // configuration, which is nothing but sealed values and so cannot go
    // without a passphrase.
    status([
      `${number.format(rows)} rows saved.`,
      doc.secrets === "passphrase"
        ? "Credentials are sealed with your passphrase."
        : "Credentials are in the file as they are — keep it somewhere you would keep a password.",
      ...(doc.notes || []),
    ].join(" "));
  } catch (failure) {
    status(failure.message);
  }
}

// What came back from a restore: the counts that say it landed, and any
// warning, which is the only part somebody has to act on.
function drawReport(report) {
  const host = qs("[data-backup-report]");
  if (!host) return;
  const box = el("div", "rounded-lg border border-border bg-background/40 p-3 space-y-1");
  box.append(el("p", "text-sm",
    `${number.format(report.rows)} row${report.rows === 1 ? "" : "s"} restored from a backup taken by ${report.guard_version || "an unknown build"}.`));
  box.append(el("p", `text-xs ${muted}`,
    `${number.format(report.sealed || 0)} credential${report.sealed === 1 ? "" : "s"} re-sealed with this instance's key` +
    (report.blank ? `, ${number.format(report.blank)} left empty for you to set again.` : ".")));
  for (const warning of report.warnings || []) {
    box.append(el("p", "text-xs text-destructive", warning));
  }
  if (report.restart) {
    box.append(el("p", `text-xs ${muted}`,
      "Stored settings are read at startup, so they take a restart to come into force."));
  }
  host.replaceChildren(box);
}

async function restoreBackup() {
  const input = qs("[data-backup-file]");
  const file = input?.files?.[0];
  if (!file) {
    restoreStatus("Choose a backup file first.");
    return;
  }
  let doc;
  try {
    doc = JSON.parse(await file.text());
  } catch {
    restoreStatus("That file is not JSON, so it is not a Guard backup.");
    return;
  }
  const sections = (doc.tables || []).length;
  const rows = (doc.tables || []).reduce((total, table) => total + (table.rows?.length || 0), 0);
  const taken = doc.created_ns ? new Date(doc.created_ns / 1e6).toLocaleString() : "an unknown time";
  if (!await ask({
    title: "Replace this instance's configuration?",
    body: `${sections} section${sections === 1 ? "" : "s"} and ${number.format(rows)} row${rows === 1 ? "" : "s"}, ` +
      `taken by ${doc.guard_version || "an unknown build"} at ${taken}. ` +
      "Everything on this instance that is not in the file is removed — machines, commands, rules, secrets, members.",
    detail: doc.secrets === "passphrase"
      ? "The file carries credentials, sealed with a passphrase."
      : doc.secrets === "plaintext"
        ? "The file carries the credentials themselves; they are sealed again with this instance's key on the way in."
        : "The file carries no credentials, so every stored password and API key on this instance is cleared.",
    confirm: "Replace it",
    phrase: "replace",
  })) return;

  restoreStatus("Restoring…");
  try {
    const report = await request("/api/backup/restore", {
      method: "POST", headers: adminHeaders(),
      body: JSON.stringify({ passphrase: qs("[data-backup-restore-passphrase]")?.value || "", backup: doc }),
    });
    drawReport(report);
    restoreStatus("Restored.");
    const restart = qs("[data-backup-restart]");
    if (restart) restart.hidden = !report.restart;
    // Every page's cached copy is now about an instance that no longer exists.
    forgetAll();
    await refreshBackup();
  } catch (failure) {
    restoreStatus(failure.message);
  }
}

// The same exit the configuration page uses: guard restarts by exiting, and
// whatever supervises it brings it back against the environment it just stored.
async function restart() {
  restoreStatus("Restarting…");
  try {
    await request("/api/config/restart", { method: "POST", headers: adminHeaders() });
    restoreStatus("Restarting — this page will reconnect on its own.");
  } catch (failure) {
    restoreStatus(failure.message);
  }
}

// The file control is a button, so the press has to reach the input the browser
// will actually open a dialog for.
document.addEventListener("change", (event) => {
  const input = event.target.closest("[data-backup-file]");
  if (!input) return;
  const name = qs("[data-backup-filename]");
  if (name) name.textContent = input.files?.[0]?.name || "No file chosen";
  restoreStatus("");
});

document.addEventListener("click", (event) => {
  if (event.target.closest("[data-backup-pick]")) { qs("[data-backup-file]")?.click(); return; }
  if (event.target.closest("[data-backup-export]")) { exportBackup(); return; }
  if (event.target.closest("[data-backup-restore]")) { restoreBackup(); return; }
  if (event.target.closest("[data-backup-restart]")) { restart(); return; }
});
